### PR TITLE
refactor(runtime): document Promise-based queueing in dispatchHooks

### PR DESCRIPTION
**Pull request by alicewriteswrongs**
This documents and slightly simplifies code which uses `Promise.then` to enqueue a series of rerendering-related actions which need to take place when a component gets updated. In particular, the previous implementation used a helper function called `then` which was a) `any`-typed and b) not necessary.

This also adds JSDoc to `dispatchHooks` and some explanatory inline comments.

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [ ] Build (`npm run build`) was run locally and any changes were pushed
- [ ] Unit tests (`npm test`) were run locally and passed
- [ ] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [ ] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

GitHub Issue Number: N/A

## What is the new behavior?

This shouldn't change anything about how components are rendered, but is basically just a small simplification of a bit of code I read through recently and found confusing.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by alicewriteswrongs — COMMENTED** [review]
one little note

File: src/runtime/update-component.ts [hunk 1]
```diff
@@ -32,12 +32,35 @@ export const scheduleUpdate = (hostRef: d.HostRef, isInitialLoad: boolean) => {
...
+ */
+const dispatchHooks = (hostRef: d.HostRef, isInitialLoad: boolean): Promise<void> => {
   const elm = hostRef.$hostElement$;
   const endSchedule = createTime('scheduleUpdate', hostRef.$cmpMeta$.$tagName$);
-  const instance = BUILD.lazyLoad ? hostRef.$lazyInstance$ : (elm as any);
+  const instance = BUILD.lazyLoad ? hostRef.$lazyInstance$ : elm;
+
+  // We're going to use this variable together with {@link enqueue} to implement a
+  // little promise-based queue. We start out with it `undefined`. When we add
+  // the first function to the queue we'll set the this variable to be that
```
Comment (rwaskiewicz) [inline src/runtime/update-component.ts:52]: Non-blocking nit:
```suggestion
  // the first function to the queue we'll set this variable to be that
```

File: src/runtime/update-component.ts [hunk 2]
```diff
@@ -32,12 +32,35 @@ export const scheduleUpdate = (hostRef: d.HostRef, isInitialLoad: boolean) => {
...
+ * @returns an empty Promise which is used to enqueue a series of operations for
+ * the component
+ */
+const dispatchHooks = (hostRef: d.HostRef, isInitialLoad: boolean): Promise<void> => {
   const elm = hostRef.$hostElement$;
   const endSchedule = createTime('scheduleUpdate', hostRef.$cmpMeta$.$tagName$);
-  const instance = BUILD.lazyLoad ? hostRef.$lazyInstance$ : (elm as any);
+  const instance = BUILD.lazyLoad ? hostRef.$lazyInstance$ : elm;
+
+  // We're going to use this variable together with {@link enqueue} to implement a
```
Comment (rwaskiewicz) [inline src/runtime/update-component.ts:50]: Non-blocking nit - these JSDoc tags (`{@link}`) don't necessarily get detected in editors when using single line comments

File: src/runtime/update-component.ts [hunk 3]
```diff
@@ -48,27 +71,61 @@ const dispatchHooks = (hostRef: d.HostRef, isInitialLoad: boolean) => {
...
+
+  return enqueue(maybePromise, () => updateComponent(hostRef, instance, isInitialLoad));
 };
 
+/**
+ * This function uses a Promise to implement a simple first-in, first-out queue
+ * of functions to be called.
+ *
+ * The queue is ordered on the basis of the first argument. If it's
+ * `undefined`, then nothing is on the queue yet, so the provided funcion can be
```
Comment (rwaskiewicz) [inline src/runtime/update-component.ts:109]:
```suggestion
 * `undefined`, then nothing is on the queue yet, so the provided function can be
```

File: src/runtime/update-component.ts [hunk 4]
```diff
@@ -48,27 +71,61 @@ const dispatchHooks = (hostRef: d.HostRef, isInitialLoad: boolean) => {
...
+ * The idea is that then the return value of that enqueueing operation is kept
+ * around, so that if it was a `Promise` then subsequent functions can be
+ * enqueued by calling this function again with that `Promise` as the first
+ * argument.
+ *
+ * @param maybePromise either a `Promise` which should resolve before the next function is called or an 'empty' sentinel
+ * @param fn a function to enqueue
+ * @returns either a `Promise` or the return value of the provided function
+ */
+function enqueue(maybePromise: Promise<void> | undefined, fn: () => Promise<void>): Promise<void> | undefined {
```
Comment (rwaskiewicz) [inline src/runtime/update-component.ts:120]: Could we make this an arrow function similar to the original `then` impl please?

**Review by rwaskiewicz — COMMENTED** [review]
All in all LGTM - nothing but small nits on my end

**Review by tanner-reits — APPROVED** [review]
Looks good aside from Ryan's few comments

File: src/runtime/update-component.ts (hunk 4 above)
Comment (alicewriteswrongs) [reply src/runtime/update-component.ts:120]: Sure, we can do arrow + ternary for more compactness

**Pull request merged**